### PR TITLE
/_admin/config

### PR DESCRIPTION
**Pull request by adrw**
Refactored config class to better separate tasks of loading config.yamls & merging them into a single JSON so that the `/_admin/config` endpoint can return the relevant information for display int the admin dashboard.

Issue #272 was discovered. Temporary fix is that each module exposing data to the `/_admin` dashboard or other services will need to have a separate `{Module}WebModule` that installs the relevant web actions so that tests of the core module functionality do not require a full web server and all related dependencies.

### CODE REVIEW

File: misk/src/main/kotlin/misk/MiskModule.kt [hunk 1]
```diff
@@ -13,6 +13,8 @@ import javax.inject.Singleton
 
 class MiskModule : AbstractModule() {
   override fun configure() {
+    binder().disableCircularProxies()
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/MiskModule.kt]: nice

File: misk/src/main/kotlin/misk/config/AppName.kt [hunk 2]
```diff
@@ -2,10 +2,11 @@ package misk.config
 
 import javax.inject.Qualifier
 
+//@TODO(jwilson swankjesse) make sure same targets everywhere
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/config/AppName.kt]: remove this comment; I’ll track here: https://github.com/square/misk/issues/262

File: misk/src/main/kotlin/misk/config/ConfigWebModule.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package misk.config
+
+import misk.inject.KAbstractModule
+import misk.web.WebActionModule
+
+//    @TODO(jwilson swankjesse) https://github.com/square/misk/issues/272 hacky fix
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/config/ConfigWebModule.kt]: remove the TODO, the issue is good enough

File: misk/src/main/kotlin/misk/config/MiskConfig.kt [hunk 4]
```diff
@@ -25,35 +27,13 @@ object MiskConfig {
...
-            mapper.readerFor(JsonNode::class.java)
-          } else {
-            mapper.readerForUpdating(jsonNode)
-          }
-          jsonNode = objectReader.readValue(it)
-        }
-      } catch (e: Exception) {
-        throw IllegalStateException("could not parse $configFileName: ${e.message}", e)
-      }
+    check(!configYamls.values.all { it == null }) {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/config/MiskConfig.kt]: wanna replace “not all are null” with “any is non-null”

File: misk/src/main/kotlin/misk/config/MiskConfig.kt [hunk 5]
```diff
@@ -66,16 +46,51 @@ object MiskConfig {
     }
   }
 
+  /** @return order of map precedence taken into account */
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/config/MiskConfig.kt]: this doesn’t read like a sentence. Maybe 

```
/** 
 * Returns a JsonNode that combines the YAMLs in `configYamls`. If two nodes define the
 * same value the last one wins.
 */
```

File: misk/src/main/kotlin/misk/config/MiskConfig.kt [hunk 6]
```diff
@@ -66,16 +46,51 @@ object MiskConfig {
...
+      try {
+        result = mapper.readerForUpdating(result).readValue(value)
+      } catch (e: Exception) {
+        throw IllegalStateException("could not parse $key: ${e.message}", e)
+      }
+    }
+    return result
+  }
+
+  /**
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/config/MiskConfig.kt]: same request : full sentence?

```
/**
 * Returns a map whose keys are the names of the source Yaml files to load, and
 * whose values are the contents of those files. If a file is absent the map’s value
 * will be null.
 */
```

File: misk/src/main/kotlin/misk/web/WebModule.kt [hunk 7]
```diff
@@ -104,5 +104,9 @@ class WebModule : KAbstractModule() {
 
     // Install infrastructure support
     install(CertificatesModule())
+
+    // Bind _admin static resources to web
+    multibind<StaticResourceMapper.Entry>()
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WebModule.kt]: nice

File: misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt [hunk 8]
```diff
@@ -0,0 +1,29 @@
+package misk.web
+
+import misk.Action
+import misk.environment.Environment
+import javax.inject.Inject
+
+internal class WideOpenDevelopmentInterceptor : NetworkInterceptor {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt]: document goals of this? Maybe TODO to remove?
(or should we remove it immediately?)

File: misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  override fun intercept(chain: NetworkChain): Response<*> {
+    val response = chain.proceed(chain.request)
+    return response.copy(
+        headers = response.headers.newBuilder()
+            .add("Access-Control-Allow-Origin", "x*")
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt]: what’s with the x ?

File: misk/src/test/kotlin/misk/config/ConfigTest.kt [hunk 10]
```diff
@@ -24,6 +24,7 @@ class ConfigTest {
   val module = Modules.combine(
       ConfigModule.create("test_app", config),
       EnvironmentModule(defaultEnv)
+//    @TODO(jwilson swankjesse) https://github.com/square/misk/issues/272 breaks here
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/config/ConfigTest.kt]: nit: typically for TODO comments we indent at the right level (so `//` is beneath `Environment) and you can chose one username for me. Lets go with jwilson for this.

File: misk/src/test/kotlin/misk/tracing/TracingConfigTest.kt [hunk 11]
```diff
@@ -24,6 +27,8 @@ class TracingConfigTest {
       ConfigModule.create("test_tracing_app", config),
       TracingModule(config.tracing),
       EnvironmentModule(defaultEnv)
+//    @TODO(jwilson swankjesse) https://github.com/square/misk/issues/272 breaks here
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/tracing/TracingConfigTest.kt]: same

File: misk/web/modules/config/src/core/containers/RootContainer.tsx [hunk 12]
```diff
@@ -10,14 +11,40 @@ export namespace RootContainer {
...
 
   private handleClick = () => {
     this.setState({ isOpen: !this.state.isOpen });
   };
 
+  componentDidMount() {
+    axios
+      .get("http://0.0.0.0:8080/_admin/config")
+      .then(response => {
+        // const newState = Object.assign({}, this.state, {
```
Comment (swankjesse) [inline misk/web/modules/config/src/core/containers/RootContainer.tsx]: commented out code : remove?

**Comment by swankjesse** [comment]
Please use GitHub’s squash feature to merge this. We want Misk’s git history to only contain the coarse-grained commits.

File: misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt (hunk 9 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt]: typo 🙈

File: misk/src/main/kotlin/misk/MiskModule.kt (hunk 1 above)
Comment (ryanhall07) [reply misk/src/main/kotlin/misk/MiskModule.kt]: maybe leave a comment on why misk is opinionated about circular proxies.

File: misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt (hunk 8 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/WideOpenDevelopmentInterceptor.kt]: Might leave it until I've tested with Webpack more but likely can be removed. I've added #278

File: misk/web/modules/config/src/core/containers/RootContainer.tsx (hunk 12 above)
Comment (adrw) [reply misk/web/modules/config/src/core/containers/RootContainer.tsx]: The React implementation is half built in this PR and will be polished and code cleaned for the next one.

**Pull request merged**